src/utils/upload/fileInspector.js

```diff
@@ -0,0 +1,67 @@
+import { Gunzip, strFromU8 } from 'fflate';
+
+import techOptions from './fileUploadSpecifications';
+import readFileToBuffer from './readFileToBuffer';
+
+const Verdict = {
+  INVALID_NAME: -2,
+  INVALID_FORMAT: -1,
+  VALID_ZIPPED: 0,
+  VALID_UNZIPPED: 1,
+};
+
+const MATRIX_SIGNATURE = Buffer.from('%%MatrixMarket', 'ascii');
+const FEATURES_SIGNATURE = Buffer.from('ENS', 'ascii');
+
+const GZIP_SIGNATURE = Buffer.from([0x1f, 0x8b]);
+
+const inspectFile = async (file, technology) => {
+  // Validate a file requested for upload to the platform.
+
+  // immediately discard file if filename is not in valid set
+  const validNames = techOptions[technology].acceptedFiles;
+  if (!validNames.has(file.name)) {
+    return Verdict.INVALID_NAME;
+  }
+
+  // if name is valid, inspect first 16 bytes to validate format
+  let data = await readFileToBuffer(file.slice(0, 16));
+
+  const isGzipped = !data.slice(0, 2).compare(GZIP_SIGNATURE);
+
+  if (isGzipped) {
+    // if gzipped, decompress a small chunk to further validate contents
+    const gunzip = new Gunzip((chunk) => {
+      data = Buffer.from(chunk.slice(0, 16));
+    });
+    gunzip.push(await readFileToBuffer(file.slice(0, 128)));
+  }
+
+  const valid = isGzipped ? Verdict.VALID_ZIPPED : Verdict.VALID_UNZIPPED;
+
+  // check matrix file starts with matrix signature
+  if (file.name.startsWith('matrix')
+    && !data.slice(0, MATRIX_SIGNATURE.length).compare(MATRIX_SIGNATURE)) {
+    return valid;
+  }
+
+  // check genes file starts with Ensembl Stable ID - ENS or "ENS
+  if ((file.name.startsWith('features') || file.name.startsWith('genes'))
+      && (!data.slice(0, 3).compare(FEATURES_SIGNATURE)
+      || !data.slice(1, 4).compare(FEATURES_SIGNATURE))) {
+    return valid;
+  }
+
+  // check barcodes file starts with a 16 digit DNA sequence
+  if (file.name.startsWith('barcodes')
+      && strFromU8(data).match(/^[ACGT]+$/)) {
+    return valid;
+  }
+
+  return Verdict.INVALID_FORMAT;
+};
+
+export {
+  inspectFile,
+  Verdict,
+};
```

Comment:
Somehow it got lost, but I had a comment here proposing that instead of exporting `Verdict` and having `src/utils/upload/processUpload.js` to have to understand what each error code in `Verdict` means, you can make function `inspectFile` return the code + the mesage directly. Then print to the user the message returned by the function if there is an error directly

Reply:
Do you mean like `return [code, message]`?

Reply:
yes

Reply:
One problem with this is that we check the verdict again here https://github.com/biomage-ltd/ui/pull/436/files#diff-cf8293615aa71eb30ec9e5199c47f4cc10f13be1ffa44f2319797cd448c85bcbR216 You need to export it anyway for this

Reply:
fair enough, then maybe leave it as is